core/src/p2p.rs

```
@@ -0,0 +1,694 @@
+    pub fn update_peer_target(&self, new_target: &[PeerId]) {
+        let mut target = self.connect_peer_target.lock().unwrap();
+        target.clear();
+        target.extend_from_slice(new_target);
+        if let Some(index) = target
+            .iter()
+            .position(|peer_id| peer_id.public_key == self.public_key)
+        {
+            target.remove(index);
+        }
+        let target = target.clone();
+        let mut connected_to_peers = self.connected_to_peers.lock().unwrap();
+        let mut to_disconnect_keys = Vec::new();
+        for public_key in connected_to_peers.keys() {
+            if target.iter().position(|id| id.public_key == *public_key).is_none() {
+                to_disconnect_keys.push(public_key.clone());
+            }
+        }
+        for key in to_disconnect_keys {
+            println!("Disconnected because not in target from {}.", key);
+            connected_to_peers.remove(&key);
+        }
+    }
+}
+
+fn read_from_socket(
+    stream: &mut TcpStream,
+    crypto: &Cryptographer,
+) -> Option<Option<NetworkMessage>> {
+    if stream.write_all(&0_u32.to_le_bytes()).is_err() {
+        // This has to be done in order to detect broken pipes.
+        return None;
+    }
+
+    let mut packet_size = 0;
+    while packet_size == 0 {
+        let mut _byte = 0_u8;
+        stream.set_nonblocking(true).unwrap();
+        let byte_count = if let Ok(value) = stream.peek(std::slice::from_mut(&mut _byte)) {
+            stream.set_nonblocking(false).unwrap();
+            value
+        } else {
+            stream.set_nonblocking(false).unwrap();
+            return Some(None);
+        };
+        if byte_count == 0 {
+            return Some(None);
+        }
+        let mut packet_size_buf = [0_u8; 4];
+        stream.read_exact(&mut packet_size_buf).ok()?;
+        packet_size = u32::from_le_bytes(packet_size_buf);
+    }
+
+    let mut buf = vec![0_u8; packet_size as usize];
+    stream.read_exact(&mut buf).ok()?;
```

We are hiding errors.